src/coreclr/jit/morph.cpp

```diff
@@ -12913,6 +12913,39 @@ GenTree* Compiler::fgOptimizeAddition(GenTreeOp* add)
             DEBUG_DESTROY_NODE(add);
             return op1;
         }
+
+        // Fold (~x + GT_CNS_INT/GT_CNS_LNG) to ((GT_CNS_INT/GT_CNS_LNG - 1) - x).
+        // Ex. (~x + 2) = (1 - x)
+        if (op1->OperIs(GT_NOT) && op2->IsIntegralConst() &&
+            (op2->AsIntCon()->IconValue() > 1 || op2->AsIntCon()->IconValue() < 1))
+        {
+        
+            add->SetOper(GT_SUB);
+        
+            add->gtOp1              = op2;
+            GenTreeIntCon* constOp2 = op2->AsIntCon();
+            constOp2->SetValueTruncating(constOp2->IconValue() - 1);
+        
+            add->gtOp2 = op1->AsOp()->gtGetOp1();
+            DEBUG_DESTROY_NODE(op1);
+        
+            return add;
+        }
```

Comment:
@jakobbotsch I added this code to support compilation of `(~x + 2)` as shown here https://godbolt.org/z/TqTfq3dEq
I don't like this part tho, because it generates worse diffs and I don't like that I have to to put Const Node to the left.
[Diff with this code](https://github.com/dotnet/runtime/files/8747290/diff_summary.16.md)
[Diff without this code](https://github.com/dotnet/runtime/files/8747291/diff_summary.17.md)
What do you think of this part of code? Have I missed anything which makes the diffs worse?

Reply:
I can see that other code in the JIT leaves `-a + C` alone:
https://github.com/dotnet/runtime/blob/f8323092eb414f9fc09a262132c65e09c5427c10/src/coreclr/jit/morph.cpp#L12883-L12893

I would do the same here to avoid introducing this pattern with a constant on the left.
Besides, it is not clear that `C - x` is always preferred over `~x + C` as it requires an extra register on x86/x64. The JIT does not have mechanism for picking between two forms like this depending on register pressure.
So I think you should go back to just optimizing  `~x + 1`. You can also experiment with `~x + C => -x + (C - 1)` and see if that unlocks further optimizations (you can check occurrences in SPMI).